src/complex_fluids/CFGiesekusRelaxation.cpp

```diff
@@ -0,0 +1,153 @@
+//
+//    * Redistributions of source code must retain the above copyright notice,
+//      this list of conditions and the following disclaimer.
+//
+//    * Redistributions in binary form must reproduce the above copyright
+//      notice, this list of conditions and the following disclaimer in the
+//      documentation and/or other materials provided with the distribution.
+//
+//    * Neither the name of The University of North Carolina nor the names of
+//      its contributors may be used to endorse or promote products derived from
+//      this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+// LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+// POSSIBILITY OF SUCH DAMAGE.
+
+#include "ibamr/CFGiesekusRelaxation.h"
+
+// Namespace
+namespace IBAMR
+{
+CFGiesekusRelaxation::CFGiesekusRelaxation(const std::string& object_name, Pointer<Database> input_db)
+    : CFRelaxationOperator(object_name)
+{
+    d_lambda = input_db->getDouble("relaxation_time");
+    d_alpha = input_db->getDouble("alpha");
+    d_sqr_root = input_db->getBoolWithDefault("square_root_evolve", false);
+    d_log_conform = input_db->getBoolWithDefault("log_conform_evolve", false);
+    return;
+} // Constructor
+
+void
+CFGiesekusRelaxation::setDataOnPatch(const int data_idx,
+                                     Pointer<Variable<NDIM>> /*var*/,
+                                     Pointer<Patch<NDIM>> patch,
+                                     const double /*data_time*/,
+                                     const bool initial_time,
+                                     Pointer<PatchLevel<NDIM>> /*patch_level*/)
+{
+    const Box<NDIM>& patch_box = patch->getBox();
+    const Pointer<CartesianPatchGeometry<NDIM>> p_geom = patch->getPatchGeometry();
+    Pointer<CellData<NDIM, double>> ret_data = patch->getPatchData(data_idx);
+    Pointer<CellData<NDIM, double>> in_data = patch->getPatchData(d_W_cc_idx);
+    ret_data->fillAll(0.0);
+    if (initial_time) return;
+    const double l_inv = 1.0 / d_lambda;
+    for (CellIterator<NDIM> i(patch_box); i; i++)
+    {
+        const CellIndex<NDIM>& idx = i();
+#if (NDIM == 2)
+        double Q_xx, Q_yy, Q_xy;
+        if (d_sqr_root)
```

Comment:
Is the following code common to all formulations that use sqrt or log conformation tensor? _Might_ be worth putting it as a helper function in the base class.